pkg/storegateway/bucket_index_postings.go

```diff
@@ -19,64 +19,126 @@ import (
+			// Try with the rest of the set matchers, maybe they can match some series.
+			continue
+		} else if err != nil {
+
+			return nil, err
+		}
+		existingKeys++
+	}
+	if existingKeys == len(keys) {
+		return keys, nil
+	}
+	if existingKeys == 0 {
+		return nil, nil
+	}
+
+	filtered := make([]labels.Label, 0, existingKeys)
+	for _, k := range keys {
+		if k == (labels.Label{}) {
+			continue
+		}
+		filtered = append(filtered, k)
+	}
+	return filtered, nil
+}
+
+// toPostingGroup returns a rawPostingGroup. toPostingGroup does not guarantee that
+// they keys of each rawPostingGroup exist in the index of the block. To verify this,
+// call rawPostingGroup.prepare() with an index reader.
+// NOTE: Derived from tsdb.postingsForMatcher
+func toPostingGroup(m *labels.Matcher) rawPostingGroup {
 	if setMatches := m.SetMatches(); len(setMatches) > 0 && (m.Type == labels.MatchRegexp || m.Type == labels.MatchNotRegexp) {
 		keys := make([]labels.Label, 0, len(setMatches))
 		for _, val := range setMatches {
-			if _, err := lvr.PostingsOffset(m.Name, val); errors.Is(err, indexheader.NotFoundRangeErr) {
-				// This label name and value doesn't exist in this block, so there are 0 postings we can match.
-				// Try with the rest of the set matchers, maybe they can match some series.
-				continue
-			}
 			keys = append(keys, labels.Label{Name: m.Name, Value: val})
 		}
 		if m.Type == labels.MatchNotRegexp {
-			return newPostingGroup(true, nil, keys), nil
+			return newRawPostingGroup(true, m.Name, keys)
 		}
-		return newPostingGroup(false, keys, nil), nil
+		return newRawPostingGroup(false, m.Name, keys)
 	}
 
 	if m.Value != "" {
 		// Fast-path for equal matching.
 		// Works for every case except for `foo=""`, which is a special case, see below.
 		if m.Type == labels.MatchEqual {
-			if _, err := lvr.PostingsOffset(m.Name, m.Value); errors.Is(err, indexheader.NotFoundRangeErr) {
-				// This label name or value doesn't exist in this block, so there are 0 postings we can match.
-				return newPostingGroup(false, nil, nil), nil
-			}
-			return newPostingGroup(false, []labels.Label{{Name: m.Name, Value: m.Value}}, nil), nil
+			return newRawPostingGroup(false, m.Name, []labels.Label{{Name: m.Name, Value: m.Value}})
 		}
 
 		// If matcher is `label!="foo"`, we select an empty label value too,
 		// i.e., series that don't have this label.
 		// So this matcher selects all series in the storage,
 		// except for the ones that do have `label="foo"`
 		if m.Type == labels.MatchNotEqual {
-			if _, err := lvr.PostingsOffset(m.Name, m.Value); errors.Is(err, indexheader.NotFoundRangeErr) {
-				// This label name or value doesn't exist in this block, so we should include all postings
-				// because all series match m.
-				return newPostingGroup(true, nil, nil), nil
-			}
-			return newPostingGroup(true, nil, []labels.Label{{Name: m.Name, Value: m.Value}}), nil
+			return newRawPostingGroup(true, m.Name, []labels.Label{{Name: m.Name, Value: m.Value}})
```

Comment:
[opinion, loosely held] At the cost of few more functions, I think would be clear to read `newRawMatchPostingGroup()` and `newRawNotMatchPostingGroup()` instead of passing `isSubtract=true/false`.

Reply:
Alternatively, have we considered having something like `type MatchOperation` whose values could be `Match` `NotMatch` (or something similar) instead of passing around a bool?

Reply:
> [opinion, loosely held] At the cost of few more functions, I think would be clear to read newRawMatchPostingGroup() and newRawNotMatchPostingGroup() instead of passing isSubtract=true/false

sure 👍 

> Alternatively, have we considered having something like type MatchOperation whose values could be Match NotMatch (or something similar) instead of passing around a bool?

if we have the underlying value of `MatchOperation` be bool, then we're effectively doing the same thing

Reply:
> [opinion, loosely held] At the cost of few more functions, I think would be clear to read newRawMatchPostingGroup() and newRawNotMatchPostingGroup() instead of passing isSubtract=true/false

added in 05b7e675f7b582196983b7d60818a6be8171d99f

> > Alternatively, have we considered having something like type MatchOperation whose values could be Match NotMatch (or something similar) instead of passing around a bool?
> 
> if we have the underlying value of `MatchOperation` be bool, then we're effectively doing the same thing

but also at this stage we're more working with sets of posting and less with matchers.